File: String/MinimumWindowSubstring/MinimumWindowSubstring.py
# 76. Minimum Window Substring
# https://leetcode.com/problems/minimum-window-substring/

"""
투포인트로 접근 하지만 너무 복잡하고 어려움.
당연하게도 실행 속도가 느림. 264ms, 18.48%보다 빠름.
복잡도도 정확히 계산을 못하겠음.
"""

from collections import defaultdict

# two-pointer
# 공간 최적화
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        tDict = defaultdict(int)
        for c in t:
            tDict[c] += 1

        tCount = len(tDict)

        left = 0
        right = 0
        maxLength = 100001
        res = ""
        while right < len(s):
            tDict[s[right]] -= 1
            if tDict[s[right]] == 0:
                tCount -= 1

            while tCount == 0:
                if maxLength > right - left:
                    res = s[left : right + 1]
                    maxLength = len(res)

                if tDict[s[left]] == 0:
                    tCount += 1

                tDict[s[left]] += 1
                left += 1
            right += 1

        return res


# two-pointer
class Solution2:
    def minWindow(self, s: str, t: str) -> str:
        tDict = defaultdict(int)

        for c in t:
            tDict[c] += 1

        print(tDict)

        tCount = len(tDict)
        matchCount = 0

        left = 0
        right = 0
        sDict = defaultdict(int)
        maxLength = 100001
        res = ""
        while right < len(s):
            c1 = s[right]
            sDict[c1] += 1
            if c1 in tDict and tDict[c1] == sDict[c1]:
                matchCount += 1

            if matchCount == tCount:
                while left <= right and matchCount == tCount:
                    c2 = s[left]
                    sDict[c2] -= 1
                    if c2 in tDict and sDict[c2] < tDict[c2]:
                        matchCount -= 1
                    left += 1

                if maxLength > right - left:
                    maxLength = right - left
                    res = s[left - 1 : right + 1]
            right += 1

        return res


# O(n+m)? O(mn)? 체크가 느림.
class Solution1:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""

        left = 0
        right = 0

        tDict = {}
        answer = []

        for c in t:
            if c in tDict:
                tDict[c] += 1
            else:
                tDict[c] = 1

        while s[left] not in tDict:
            left += 1
            if left >= len(s):
                return ""

        right = left

        while right < len(s):
            # print("lr ", left, right, tDict)
            if s[right] in tDict:
                tDict[s[right]] -= 1

            while self.checkDict(tDict):
                answer.append([left, right])

                tDict[s[left]] += 1

                left += 1
                while left < right and s[left] not in tDict:
                    left += 1

            right += 1

        if not answer:
            return ""

        lenAnswer = list(map(lambda x: x[1] - x[0], answer))
        pos = lenAnswer.index(min(lenAnswer))
        l = answer[pos][0]
        r = answer[pos][1] + 1

        return s[l:r]

    def checkDict(self, d: dict):
        for val in d.values():
            if val > 0:
                return False
        return True


if __name__ == "__main__":
    sol = Solution()

    print(sol.minWindow("ADOBECODEBANC", "ABC"))
    print(sol.minWindow("DDDADOBECODEBANC", "ABC"))
    print(sol.minWindow("DDDABBBBCCA", "ABC"))
    print(sol.minWindow("a", "a"))
    print(sol.minWindow("a", "aa"))
    print(sol.minWindow("aa", "aa"))
    print(sol.minWindow("a", "b"))

File: String/MinimumWindowSubstring/MinimumWindowSubstring.java
package String.MinimumWindowSubstring;
// 76. Minimum Window Substring
// https://leetcode.com/problems/minimum-window-substring/

import java.util.HashMap;
import java.util.Map;

// solution two-pointer array 4ms
// space 최적화
class Solution {
    char[] str;
    int[] target = new int[128];

    public String minWindow(String s, String t) {
        String res = "";
        int maxLength = 100001;

        str = s.toCharArray();

        int targetCount = 0;
        for (char c : t.toCharArray()) {
            if (target[c] == 0) {
                targetCount++;
            }
            target[c]++;
        }

        int left = 0;
        int right = 0;
        while (right < s.length()) {
            char c1 = str[right];
            target[c1]--;

            if (target[c1] == 0) {
                targetCount--;
            }

            if (0 == targetCount) {
                while (0 == targetCount) {
                    char c2 = str[left];
                    target[c2]++;
                    if (target[c2] > 0) {
                        targetCount++;
                    }
                    left++;
                }

                if (maxLength > (right - left)) {
                    maxLength = right - left;
                    res = s.substring(left - 1, right + 1);
                }
            }
            right++;
        }
        return res;
    }
}


// solution two-pointer array 2ms
class Solution4 {
    char[] str;
    int[] target = new int[128];

    public String minWindow(String s, String t) {
        String res = "";
        int maxLength = 100001;

        str = s.toCharArray();

        int targetCount = 0;
        for (char c : t.toCharArray()) {
            if (target[c] == 0) {
                targetCount++;
            }
            target[c]++;
        }

        int matchCount = 0;
        int left = 0;
        int right = 0;
        int[] map = new int[128];
        while (right < s.length()) {
            char c1 = str[right];
            map[c1]++;

            if (target[c1] > 0 && target[c1] == map[c1]) {
                matchCount++;
            }

            if (matchCount == targetCount) {
                while (left <= right && matchCount == targetCount) {
                    char c2 = str[left];
                    map[c2]--;
                    if (target[c2] > 0 && map[c2] < target[c2]) {
                        matchCount--;
                    }
                    left++;
                }

                if (maxLength > (right - left)) {
                    maxLength = right - left;
                    res = s.substring(left - 1, right + 1);
                }
            }
            right++;
        }
        return res;
    }
}


// solution two-pointer hashmap 26ms
class Solution3 {
    char[] str;
    Map<Character, Integer> target = new HashMap<>();

    public String minWindow(String s, String t) {
        String res = "";
        int maxLength = 100001;

        str = s.toCharArray();

        for (char c : t.toCharArray()) {
            target.put(c, target.getOrDefault(c, 0) + 1);
        }

        int matchCount = 0;
        int left = 0;
        int right = 0;
        Map<Character, Integer> map = new HashMap<>();
        while (right < s.length()) {
            char c1 = str[right];
            map.put(c1, map.getOrDefault(c1, 0) + 1);

            // if (target.containsKey(c1) && map.get(c1) == target.get(c1)) { //ERROR 왜?
            if (target.containsKey(c1) && map.get(c1).equals(target.get(c1))) {
                matchCount++;
            }

            if (matchCount == target.size()) {
                // System.out.println(map);
                while (left <= right && matchCount == target.size()) {
                    char c2 = str[left];
                    map.put(c2, map.getOrDefault(c2, 0) - 1);
                    if (target.containsKey(c2) && map.get(c2) < target.getOrDefault(c2, 0)) {
                        matchCount--;
                    }
                    left++;
                }

                if (maxLength > (right - left)) {
                    maxLength = right - left;
                    res = s.substring(left - 1, right + 1);
                }
            }
            right++;
        }
        return res;
    }
}


// two-pointer
// check가 너무 느림. 89ms
class Solution2 {
    char[] str;
    Map<Character, Integer> target = new HashMap<>();

    public String minWindow(String s, String t) {
        str = s.toCharArray();

        for (char c : t.toCharArray()) {
            target.put(c, target.getOrDefault(c, 0) + 1);
        }

        Map<Character, Integer> map = new HashMap<>();

        int left = 0;
        int right = 0;
        String res = "";
        int maxLength = 100001;
        while (right < s.length()) {
            map.put(str[right], map.getOrDefault(str[right], 0) + 1);
            if (check(map)) {
                System.out.println(map);
                while (left <= right) {
                    map.put(str[left], map.getOrDefault(str[left], 0) - 1);
                    left++;
                    if (map.get(str[left - 1]) < target.getOrDefault(str[left - 1], 0)) {
                        break;
                    }
                }
                if (maxLength > (right - left)) {
                    maxLength = right - left;
                    res = s.substring(left - 1, right + 1);
                }
            }
            right++;
        }
        return res;
    }

    private boolean check(Map<Character, Integer> map) {
        for (Map.Entry<Character, Integer> entry : target.entrySet()) {
            if (!map.containsKey(entry.getKey()) || map.get(entry.getKey()) < entry.getValue()) {
                return false;
            }
        }
        return true;
    }
}


// brute-force + 약간의 dp?
// 모든 윈도우 사이즈별로 체크
class Solution1 {
    char[] str;
    Map<Character, Integer> target = new HashMap<>();

    public String minWindow(String s, String t) {
        int m = t.length();
        str = s.toCharArray();

        for (char c : t.toCharArray()) {
            target.put(c, target.getOrDefault(c, 0) + 1);
        }

        for (int i = m; i <= s.length(); ++i) {
            // System.out.println(i);
            Map<Character, Integer> map = new HashMap<>();
            for (int j = 0; j < i; ++j) {
                map.put(str[j], map.getOrDefault(str[j], 0) + 1);
            }

            for (int j = 0; j < s.length() - i + 1; ++j) {
                System.out.println(map);
                if (check(map)) {
                    return s.substring(j, j + i);
                }
                map.put(str[j], map.getOrDefault(str[j], 0) - 1);
                if (j + i >= s.length()) {
                    break;
                }
                map.put(str[j + i], map.getOrDefault(str[j + i], 0) + 1);
            }
        }
        return "";
    }

    private boolean check(Map<Character, Integer> map) {
        for (Map.Entry<Character, Integer> entry : target.entrySet()) {
            if (!map.containsKey(entry.getKey()) || map.get(entry.getKey()) < entry.getValue()) {
                return false;
            }
        }
        return true;
    }
}


public class MinimumWindowSubstring {
    public static void main(String[] args) {
        testSol("BANC", "ABC", "BANC");
        testSol("ADOBECODEBANC", "ABC", "BANC");
        testSol("a", "a", "a");
        testSol("a", "aa", "");
        testSol("bbaa", "aba", "baa");
    }

    static void testSol(String input1, String input2, String output) {
        Solution sol = new Solution();
        // todo : solution match
        String res = sol.minWindow(input1, input2);
        if (res.equals(output)) {
            System.out.println("O : " + res);
        } else {
            System.out.println("x : " + res + "	expect : " + output);
        }
    }
}
